### PR TITLE
Fix 4099 by opening event subscriptions after start event's output mappings were applied

### LINKED ISSUE #4099: A workflow with start event message will deploy but fail to start if it has an event sub process
**Describe the bug**
A workflow that is started with a start event message will deploy but fail to start if it has an event sub process (this is not the same as the deploy time issue caused by the modeller leaving out correlationkey ).

Variables added to the process start message don't seem to visible when the broker tries to subscribe the event sub process to its start message so it it complains it can't see the correlation key (it looks for the right variable but can't find it).

This also sometimes causes the operate UI to get in bad state where it always reports the incident but can't always show the instance details.

Workaround: Placing the event subprocess within a wrapping subprocess fixes the issue.

**To Reproduce**
Steps to reproduce the behavior (using attached files)
zbctl --insecure deploy correlationIssue.bpmn 
zbctl --insecure deploy correlationIssueWorkaround.bpmn

(escape as needed for your shell)
zbctl --insecure publish message myStartMessage --variables '{"orderID":123}' --correlationKey 123

observe behaviour in operate UI
Process_CorrelationIssue fails with an incident (operate may or may not let you inspect it)
Process_CorrelationIssueWorkaround successfully starts

**Expected behavior**
both workflows start without error

**Log/Stacktrace**
Process_CorrelationIssue Failed to extract the correlation-key by 'orderID': no value found

**Environment:**
 - OS: Linux

[correlationIssue.bpmn.txt](https://github.com/zeebe-io/zeebe/files/4360758/correlationIssue.bpmn.txt)
issue

[correlationIssueWorkaround.bpmn.txt](https://github.com/zeebe-io/zeebe/files/4360759/correlationIssueWorkaround.bpmn.txt)
workaround

 - Zeebe Version: 0.22.1
 - Configuration: simple setup based on standard helm chart with reduced resources

### PR DESCRIPTION
**Pull request by pihme**
## Description

This moves the step when event subscriptions are opened from `ProcessProcessor` and `SubprocessProcessor` to `StartEventProcessor`. This way, subscriptions and the values used for them will only be evaluated after output mappings of the start event have been applied.

This fits better to user expectations and fixes #4099 

## Related issues

closes #4099

## Definition of Done
Code changes:
* [X] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [X] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

### CODE REVIEW

**Comment by pihme** [comment]
Should be merged into `stable/8.0` after it has been created

**Review by korthout — CHANGES_REQUESTED** [review]
Nice work @pihme 👍 

❌ The test case shows clearly that the bug has been resolved. I think the only thing that's missing is a test case that shows that the output mappings of a start event are applied before the subscriptions are opened. Since this is the behavior that changes for users, I'd like to have a test for it.

🔧 Next to that, I just have a few nitpick suggestions.

Lastly, please open a bug that output mappings of a none start event don't allow to modify how the variables are merged into the process. This is the bug we discovered when we played around with none start event.

File: engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/event/StartEventProcessor.java [hunk 1]
```diff
@@ -40,8 +47,21 @@ public void onActivate(final ExecutableStartEvent element, final BpmnElementCont
 
   @Override
   public void onComplete(final ExecutableStartEvent element, final BpmnElementContext context) {
+    final ExecutableCatchEventSupplier flowScope =
+        (ExecutableCatchEventSupplier) element.getFlowScope();
```
[the remark this answers is not in the document]
Comment (pihme) [reply engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/event/StartEventProcessor.java:51]: I think it's fine

File: engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/event/StartEventProcessor.java [hunk 2]
```diff
@@ -40,8 +47,21 @@ public void onActivate(final ExecutableStartEvent element, final BpmnElementCont
...
     variableMappingBehavior
         .applyOutputMappings(context, element)
+        .flatMap(
+            ok ->
+                eventSubscriptionBehavior.subscribeToEvents(
+                    flowScope,
+                    context.copy(
+                        flowScopeInstance.getKey(),
+                        flowScopeInstance.getValue(),
+                        flowScopeInstance.getState())))
```
[the remark this answers is not in the document]
Comment (pihme) [reply engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/event/StartEventProcessor.java]: Not done exactly as you suggested, but in the spirit of what you suggested

**Comment by pihme** [comment]
> Lastly, please open a bug that output mappings of a none start event don't allow to modify how the variables are merged into the process. This is the bug we discovered when we played around with none start event.

No longer necessary. The fix solved this. I added another test with none start event. Please double check.

**Review by korthout — APPROVED** [review]
Thanks for the changes @pihme LGTM 👍 

Note that the additional bug we discovered is not fixed by these changes. If you create a process instance with some variables and the none start event has output mappings, then variables that were passed in are still copied into the process instance. This shouldn't happen as the output mappings should allow to completely redefine how and which variables are merged into the process instance. Please open an issue for it. If you want we can also discuss it F2F.

File: engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/event/StartEventProcessor.java [hunk 3]
```diff
@@ -75,4 +70,13 @@ public void onTerminate(final ExecutableStartEvent element, final BpmnElementCon
...
   }
+
+  private BpmnElementContextImpl buildContextForFlowScopeInstance(
+      final BpmnElementContext context) {
+    final var flowScopeInstance = stateBehavior.getFlowScopeInstance(context);
+    final var flowScopeInstanceContext = new BpmnElementContextImpl();
+    flowScopeInstanceContext.init(
+        flowScopeInstance.getKey(), flowScopeInstance.getValue(), flowScopeInstance.getState());
+    return flowScopeInstanceContext;
+  }
```
[the remark this answers is not in the document]
Comment (pihme) [reply engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/event/StartEventProcessor.java:81]: Couple of reasons: 
* Keeping the change as small as possible for backporting
* There already is an `init(...)` method which screams like it should just take an element instance. So probably somebody had thought about it and decided against it
* Overall, I don't understand the pattern of empty constructor and init method and I feel that adding a convenience constructor would jeopardize the intended design. In particular, it seems deliberate to me that this class has no dependency to `ElementInstance` 
* In case of doubt, I aim to use what is there instead of making changes to it

**Comment by pihme** [comment]
Bug created: https://github.com/camunda/zeebe/issues/9061

**Comment by pihme** [comment]
bors merge

**Comment by pihme** [comment]
bors merge
